File: moviejava/app/src/main/java/com/example/moviejava/ui/film/FilmViewModel.java
package com.example.moviejava.ui.film;

import android.os.Build;

import androidx.annotation.RequiresApi;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.moviejava.data.local.dao.AppDao;
import com.example.moviejava.data.local.entities.MovieEntity;
import com.example.moviejava.data.remote.film.FilmRepository;
import com.example.moviejava.model.State;
import com.example.moviejava.model.casts.CastsFilm;
import com.example.moviejava.model.film.Film;
import com.example.moviejava.model.film.video.Video;
import com.example.moviejava.ui.base.BaseViewModel;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.schedulers.Schedulers;

@HiltViewModel
public class FilmViewModel extends BaseViewModel {
    private FilmRepository repository;
    private AppDao dao;

    @Inject
    public FilmViewModel(FilmRepository repository, AppDao dao) {
        this.repository = repository;
        this.dao = dao;

    }

    private final MutableLiveData<State<Film>> _film = new MutableLiveData<>();
    public LiveData<State<Film>> film = _film;

    private final MutableLiveData<State<Video>> _video = new MutableLiveData<>();
    public LiveData<State<Video>> video = _video;

    private final MutableLiveData<State<CastsFilm>> _cast = new MutableLiveData<>();
    public LiveData<State<CastsFilm>> cast = _cast;

    @RequiresApi(api = Build.VERSION_CODES.N)
    public void loadData(int id) {
        launchRxSingle(repository.getFilm(id),
                result -> _film.postValue(State.success(result)),
                null);
        launchRxSingle(
             repository.getVideo(id),
                result -> _video.postValue(State.success(result)),
                null);
        launchRxSingle(
               repository.getCast(id),
                result -> _cast.postValue(State.success(result)),
                null);
//        repository.getFilm(id).observeOn(AndroidSchedulers.mainThread())
//                .subscribe(new Observer<Film>() {
//                    @Override
//                    public void onSubscribe(@NonNull Disposable d) {
//
//                    }
//
//                    @Override
//                    public void onNext(@NonNull Film film) {
//                        _film.setValue(film);
//                    }
//
//                    @Override
//                    public void onError(@NonNull Throwable e) {
//                        e.printStackTrace();
//                    }
//
//                    @Override
//                    public void onComplete() {
//                        Log.d("complete", "not error");
//                    }
//                });
//        repository.getVideo(id).observeOn(AndroidSchedulers.mainThread())
//                .subscribe(new Observer<Video>() {
//                    @Override
//                    public void onSubscribe(@NonNull Disposable d) {
//
//                    }
//
//                    @Override
//                    public void onNext(@NonNull Video video) {
//                        _video.setValue(video);
//                    }
//
//                    @Override
//                    public void onError(@NonNull Throwable e) {
//                        e.printStackTrace();
//                    }
//
//                    @Override
//                    public void onComplete() {
//                        Log.d("complete", "not error");
//                    }
//                });
//        repository.getCast(id).observeOn(AndroidSchedulers.mainThread())
//                .subscribe(new Observer<CastsFilm>() {
//                    @Override
//                    public void onSubscribe(@NonNull Disposable d) {
//
//                    }
//
//                    @Override
//                    public void onNext(@NonNull CastsFilm castsFilm) {
//                        _cast.setValue(castsFilm);
//                    }
//
//                    @Override
//                    public void onError(@NonNull Throwable e) {
//                        e.printStackTrace();
//                    }
//
//                    @Override
//                    public void onComplete() {
//                        Log.d("complete", "not error");
//                    }
//                });
    }

    public void addToDb(Film f) {
        MovieEntity m = new MovieEntity();
        m.id = f.id;
        m.adult = f.adult;
        m.backdropPath = f.backdropPath;
        m.posterPath = f.posterPath;
        m.releaseDate = f.releaseDate;
        m.title = f.title;
        m.rate = f.voteAverage;
        Completable.fromRunnable(() -> dao.addRecord(m)).subscribeOn(Schedulers.io()).subscribe();

    }


}

File: movie/app/src/main/java/com/example/movie/ui/film/FilmViewModel.kt
package com.example.movie.ui.film

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.example.movie.data.local.dao.AppDao
import com.example.movie.data.local.entities.MovieEntity
import com.example.movie.data.remote.film.FilmRepository
import com.example.movie.model.State
import com.example.movie.model.casts.CastsFilm
import com.example.movie.model.film.Film
import com.example.movie.model.film.video.Video
import com.example.movie.ui.base.BaseViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class FilmViewModel @Inject constructor(
    private val repository: FilmRepository, private val dao: AppDao
) : BaseViewModel() {
    private val _film = MutableLiveData<State<Film>>()
    val film: MutableLiveData<State<Film>>
        get() = _film

    private val _video = MutableLiveData<State<Video>>()
    val video: LiveData<State<Video>>
        get() = _video
    private val _casts = MutableLiveData<State<CastsFilm>>()
    val casts: LiveData<State<CastsFilm>>
        get() = _casts

    fun loadData(id: Int) {
        launchViewModel(repository.getFilm(id), {
            _film.postValue(State.success())
        }, {
            Log.e("err", it.message.toString())
        })

        launchViewModel(repository.getVideo(id), {
            _video.postValue(State.success(it))
        }, {
            Log.e("err", it.message.toString())
        })

        launchViewModel(repository.getCastFilm(id), {
            _casts.postValue(State.success(it))
        }, {
            Log.e("err", it.message.toString())
        })
    }

    fun addToDb(r: Film) {
        val m = MovieEntity(
            id = r.id,
            adult = r.adult,
            backdropPath = r.backdropPath,
            posterPath = r.posterPath,
            releaseDate = r.releaseDate,
            title = r.title,
            rate = r.voteAverage
        )
        viewModelScope.launch(Dispatchers.IO) {
            dao.addRecord(m)
        }
    }

}